Issue #3207: ASOF JOIN Compilation

hawkfish opened the pull request:
Part 1 of the ASOF JOIN implementation. 

This PR adds ASOF to the grammar and converts it into an inequality join on a window:

```sql
SELECT p.key, p.ts, e.value 
FROM probe p ASOF JOIN event e
  ON p.key = e.key AND p.ts >= e.ts

-- Translates to

SELECT p.key, p.ts, e.value
FROM probe p LEFT JOIN (
  SELECT value, ts, LEAD(ts, 1, 'infinity') OVER (PARTITION BY key ORDER BY ts) AS ts_temp
  ) e
  ON p.key = e.key AND p.ts >= e.ts AND p.ts < e.ts_temp
```

where the left join is implemented using the `IEJoin` operator.

This is not very efficient (it does three sorts), and part 2 will add a new physical operator that uses only one partitioned sort and the outer partitioning from hash join. But it will hopefully make it easier to specify these joins in a future-proof manner.

Commit:
Issue #3207: ASOF Join Grammar

Checkpoint commit with the ASOF grammar additions.

Commit:
Issue #3207: ASOF Join Planning

Checkpoint commit.

Commit:
Merge branch 'master' into asof-joins

Commit:
Issue #3207: ASOF Join Planning

Implement ASOF joins using IEJoin + Window.

Commit:
Issue #3207: ASOF Equality Clauses

Fix and test equality clauses in ASOF Joins.

Commit:
Merge branch 'master' into asof-joins

Commit:
Issue #3207: ASOF JOIN Serialisation

Add missing EnumSerializer::StringToEnum case.

Mytherin reviewed (commented):
Thanks for the PR! Looks good. Some comments:

hawkfish commented on src/execution/physical_plan/plan_comparison_join.cpp:
```diff
@@ -200,6 +202,111 @@ unique_ptr<PhysicalOperator> PhysicalPlanGenerator::CreatePlan(LogicalComparison
+			}
+		}
+
+		auto proj =
```
I assumed we needed to strip the `asof_temp` column?

hawkfish commented on src/common/serializer/enum_serializer.cpp:
```diff
@@ -303,6 +305,8 @@ const char *EnumSerializer::EnumToString(JoinType value) {
 	case JoinType::MARK:
 		return "MARK";
+	case JoinType::ASOF:
+		return "ASOF";
```
That's where I started, but it has conditions so it's a lot more like `REGULAR`. I also toyed with converting `isNatural` to an `enum`.

Commit:
Merge branch 'master' into asof-joins

Commit:
Issue #3207: ASOF Projection Refactor

Move the join projection logic into PhysicalProjectionOperator.

Commit:
Issue #3207: ASOF Projection Refactor

Fix tidy include.

Commit:
Merge branch 'master' into asof-joins

Commit:
Issue #3207: ASOF Outer Joins

* Decouple the JoinType from the JoinRefType.
* Convert ASOF to be a JoinRefType.
* Attach a JoinRefType to LogicalConditionalJoin to indicate how the conditions should be interpreted
* Stop the optimiser from thinking it knows how to deal with ASOF joins.

Commit:
Merge branch 'asof-joins' of github.com:hawkfish/duckdb into asof-joins

Commit:
Issue #3207: ASOF Outer Joins

Regenerate plans file.

Commit:
Issue #3207: ASOF Outer Joins

Fix copy elision.

Mytherin commented on test/sql/join/asof/test_asof_join.test:
```diff
@@ -0,0 +1,352 @@
+		LEAD(begin, 1, 'infinity'::DOUBLE) OVER (ORDER BY begin ASC) AS end
+	FROM events0
+) e
+ON p.ts >= e.begin AND p.ts < e.end
```
Could we add some extra tests:

* Join on a timestamp range
* Join on a string range (it is fine if that throws a non-internal exception)
* Join on integer ranges (it is fine if that throws a non-internal exception)
* Multiple ranges in one join condition
* What if we invert the comparisons (e.g. `e.begin <= p.ts AND e.end > p.ts`)
* What if we have a different closed range (e.g. `p.ts > e.begin AND p.ts <= e.end`)
* What if we have a range that is not actually a range (e.g. `p.ts >= e.begin and p.other_col < e.end`)
* ASOF join on a constant (e.g. `ON 1=1`)
* Could we add some tests where we have NULL values in the tables?
* Could we add some tests where we have infinity values (e.g. `NaN`, or `infinity::timestamp`) in the tables?
* What happens if we use an ASOF join inside of a correlated subquery? (it is fine if that throws an exception)

hawkfish replied:
Multiple ranges was covered by the last error test, and you can't have invalid ranges because the bounds are implied (i.e., `end` is synthesised). But I'll have a go at the rest.

hawkfish replied:
Got all of these except the last one. How would you write an ASOF join inside a correlated subquery? Aren't the joins in correlated subqueries generated by the planner?

Mytherin replied:
Ah, I meant an ASOF join inside a subquery that has correlated columns in the left and/or right side, see e.g. [here](https://github.com/duckdb/duckdb/blob/master/test/sql/subquery/scalar/test_subquery_any_join.test#L4)

Mytherin commented on src/execution/physical_plan/plan_comparison_join.cpp:
```diff
@@ -200,6 +202,82 @@ unique_ptr<PhysicalOperator> PhysicalPlanGenerator::CreatePlan(LogicalComparison
 
+	if (op.join_reftype == JoinRefType::ASOF) {
+		//	Validate
+		vector<idx_t> equi_indexes;
```
I wonder if it makes sense for the ASOF join to be a separate `LogicalOperator` similar to the `LogicalPositionalJoin` - e.g. a `LogicalASOFJoin`. It seems like the ASOF join code is mostly disjoint from the rest of the join code, and merging it into the `LogicalComparisonJoin` requires taking extra care during optimizations and such.

hawkfish replied:
Yeah I thought about that and decided not to for some reason. But it was earlier in the process so I should probably revisit it.

hawkfish replied:
It touched a lot of code, but it's probably the Right Thing To Do.

Mytherin reviewed (approved):
Thanks for the changes! Looks great. Some additional comments:

Commit:
Merge branch 'master' into asof-joins

Commit:
Issue #3207: ASOF Logical Joins

Switch to using a separate LogicalOperator to model AsOf joins.

Commit:
Issue #3207: ASOF Joins Tests

Beef up the test suite with multiple data types and corner cases.
Fix arbitrary expression planning issue.

Commit:
Issue #3207: Correlated ASOF Joins

Add test to show it works.

Mytherin commented:
Thanks for the fixes! Looks great.
